### DISCUSSION TITLE
SARIF: add contextRegion on sarif report output for better readability.

### DISCUSSION THREAD

**Discussion by relzhong** [oxsecurity/megalinter#4222] [General]
Linters output sarif report often not include contextRegion info. Such as [pmd](https://github.com/pmd/pmd/blob/f0daed58744980caff3c5e1f37d9859f33a8a8b7/docs/report-examples/pmd-report.sarif.json#L68), [checkstyle](https://github.com/checkstyle/checkstyle/blob/d7c12dad6929787d36439dbb2da65763d28ab3e0/src/main/resources/com/puppycrawl/tools/checkstyle/sarif/ResultLineOnly.template#L9). so is possible for megalinter to append contentRegion information at the collect report phase? 

If we want to add contextRegion info for those linter. we can add contextRegion info directly in those project. but it seems we had to do lots of pr to those project. If we add contextRegion in megalinter, we just read source code and append contentRegion to finally report.

**Comment by nvuillam** [comment]
Why not but ... what it contextRegion ?
Can we calculate it for sure during post-processing, using only available sarif data ?

**Reply by relzhong** [reply]
[ContextRegion](https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/sarif-v2.1.0-errata01-os-complete.html#_Toc141790941:~:text=the%20entire%20artifact.-,3.29.5,-contextRegion%20property) define in sarif specification.  
We can't calculate from existed sarif data as other linter not fill the information.  but we can access it directory from source code.

**Reply by nvuillam** [reply]
I think it's the role of source linters to generate complete SARIF output, i'm not sure that we can calculate it from Megalinter, which is just an aggregator :/

If you have ideas about how to calculate that from MegaLinter, you can try a PR of course !

Otherwise, I think the best is to request source linter authors to complete their SARIF output so it contains ContextRegion

**Reply by llaville** [reply]
Totally agree here with @nvuillam. 
@relzhong if your ecosystem is PHP, I suggest to use my project https://github.com/llaville/sarif-php-sdk that support full implementation of SARIF 2.1.0

BTW, PHP Linters included with MegaLinter already use this SDK and provide SARIF support (see https://megalinter.io/latest/descriptors/php/)

**Reply by relzhong** [reply]
Will try to contact source linter to add this feature, I think is better solution, thanks.
